application/src/main/java/org/togetherjava/tjbot/commands/github/GitHubCommand.java

```diff
@@ -0,0 +1,122 @@
+
+    private Instant lastCacheUpdate;
+    private List<String> autocompleteGHIssueCache;
+
+    public GitHubCommand(GitHubReference reference) {
+        super("github-search", "Search configured GitHub repositories for an issue/pull request",
+                CommandVisibility.GUILD);
+
+        this.reference = reference;
+
+        getData().addOption(OptionType.STRING, TITLE_OPTION,
+                "Title of the issue you're looking for", true, true);
+
+        updateCache();
+    }
+
+    @Override
+    public void onSlashCommand(SlashCommandInteractionEvent event) {
+        String title = event.getOption(TITLE_OPTION).getAsString();
+        Matcher matcher = GitHubReference.ISSUE_REFERENCE_PATTERN.matcher(title);
+
+        if (!matcher.find()) {
+            event.reply(
+                    "Could not parse your query. Was not able to find an issue number in it (e.g. #207).")
+                .setEphemeral(true)
+                .queue();
+
+            return;
+        }
+
+        int issueId = Integer.parseInt(matcher.group(GitHubReference.ID_GROUP));
+        reference.findIssue(issueId)
+            .ifPresentOrElse(issue -> event.replyEmbeds(reference.generateReply(issue)).queue(),
+                    () -> event.reply("Could not find the issue you are looking for.")
+                        .setEphemeral(true)
+                        .queue());
+    }
+
+    @Override
+    public void onAutoComplete(CommandAutoCompleteInteractionEvent event) {
+        String title = event.getOption(TITLE_OPTION).getAsString();
+
+        if (title.isEmpty()) {
+            event.replyChoiceStrings(autocompleteGHIssueCache.stream().limit(25).toList()).queue();
+        } else {
+            Queue<String> closestSuggestions =
+                    new PriorityQueue<>(Comparator.comparingInt(suggestionScorer(title)));
+
+            closestSuggestions.addAll(autocompleteGHIssueCache);
+
+            List<String> choices = Stream.generate(closestSuggestions::poll).limit(25).toList();
+            event.replyChoiceStrings(choices).queue();
+        }
+
+        if (lastCacheUpdate.isAfter(Instant.now().minus(CACHE_EXPIRES_AFTER))) {
+            updateCache();
+        }
+    }
+
+    private ToIntFunction<String> suggestionScorer(String title) {
+        // Remove the ID [#123] and then match
+        return s -> StringDistances.editDistance(title, s.replaceFirst("\\[#\\d+] ", ""));
+    }
+
+    private void updateCache() {
+        autocompleteGHIssueCache = reference.getRepositories().parallelStream().map(repo -> {
+            try {
+                return repo.getIssues(GHIssueState.ALL);
+            } catch (IOException ex) {
+                throw new UncheckedIOException(ex);
+            }
+        })
+            .flatMap(List::stream)
+            .sorted(GITHUB_ISSUE_TIME_COMPARATOR)
+            .map(issue -> "[#%d] %s".formatted(issue.getNumber(), issue.getTitle()))
+            .toList();
```

Comment:
what happened to this formatting?

Reply:
![image](https://github.com/Together-Java/TJ-Bot/assets/88111627/afff94f4-8b08-424a-ab7d-aa1e0b47d0e2)

Reply:
what's wrong?

Reply:
The formatting is weird, and why is it using parallelStream?

Reply:
sequence of repositories shouldn't be a problem, prolly just a preference by original contributor. We're following google style sheet for formatting, everything is formatted accordingly otherwise it wouldn't pass pre merge test cases.

Reply:
ok I understand the formatting, but the `#parallelStream` is probably slower, why using it then?

Reply:
from my understanding, parallel Streams are faster since we're performing each operation parallely although the difference here won't matter much since dataset is not significant enough. But it comes at a cost of sequence, so if you need order to be maintained then you should stick to normal ones.

Reply:
The problem is also that using parallel streams adds overhead for managing multiple threads, memory locality, ...
See [Baeldung: When to use parallel stream](https://www.baeldung.com/java-when-to-use-parallel-stream)
especially the NQ model, do we have more than 10000 elements in the stream?
> Parallelism can bring performance benefits in certain use cases. But parallel streams cannot be considered as a magical performance booster.

In this case, I think sequential stream is better

Reply:
fair, there's no justification for going beyond the normal route. I'll swap it with sequential stream.